README.md
# About
In this package you will find a series of functions for soil physics data analysis. These functions includes five models of water retention curve, seven methods of soil precompression stress, least limiting water range (LLWR), Integral Water Capacity (IWC), soil penetration resistance curve by Busscher's model, calculation of Soil Aggregate-Size Distribution, S Index, critical soil moisture and maximum bulk density using data from Proctor test, calculation of equivalent pore radius as a function of soil water tension, simulation of sedimentation time of soil particles through Stokes' law, simulation of soil pore size distribution, calculation of the hydraulic cut-off introduced by  Dexter et al. (2008) and simulation of soil compaction induced by agricultural field traffic. Other utilities like functions to calculate the void ratio and to determine the maximum curvature point are also available. 

# Instalation

You can install and load the released version of soilphysics from GitHub with:

```r
install.packages("devtools")

devtools::install_github("arsilva87/soilphysics")

library(soilphysics) 
```

## Soil compaction tools

Using the function stressTraffic, it it possible calculate the contact area, stress distribuition and stress propagation based on the SoilFlex model. 

```r
# Usage
stress <- stressTraffic(inflation.pressure=200, 
	   recommended.pressure=200, 
	   tyre.diameter=1.8, 
	   tyre.width=0.4, 
	   wheel.load=4000, 
	   conc.factor=c(4,5,5,5,5,5),
           layers=c(0.05,0.1,0.3,0.5,0.7,1), 
	   plot.contact.area = TRUE)

# Results
 ---------- Boundaries of Contact Area

    Parameters  Value Units
1   Max Stress 289.00 kPa  
2 Contact Area   0.29 m^2  
3  Area Length   0.83  m   
4   Area Width   0.40  m   

 ---------- Wheel Loads

          Parameters Loads (kg)
1 Applied Wheel Load       4000
2 Modeled Wheel Load       4030
3          Diference        -30

 ---------- Stress Propagation

  Layers (m) Zstress   p
1       0.05     275 169
2       0.10     262 132
3       0.30     154  62
4       0.50      85  31
5       0.70      51  18
6       1.00      28  10

 
```
![](<man/figures/contactarea.jpeg>)

Unsing the funtion soilDeformation, it is possible calculates the bulk density variation as a function of the applied mean normal stress using critical state theory, by O'Sullivan and Robertson (1996). 

```r

# Usage
soilDeformation(stress = 300,
                p.density = 2.67,
	        iBD = 1.55, 
                N = 1.9392, 
                CI = 0.06037, 
                k = 0.00608, 
                k2 = 0.01916,
                m = 1.3,graph=TRUE,ylim=c(1.4,2.0))
		
# Results
   iBD    fBD     vi     vf   I%
1 1.55 1.6385 1.7226 1.6295 5.71

```

![](<man/figures/SMP.jpeg>)


## Critical soil moisture and maximum bulk density (Proctor test)

```r
mois <- c(0.083, 0.092, 0.108, 0.126, 0.135)
bulk <- c(1.86, 1.92, 1.95, 1.90, 1.87)

# Usage
criticalmoisture(theta = mois, Bd = bulk)

# Results
          Critical Moisture and Maximum Bulk Density 

                  Sample 1
Intercept        0.4825950
mois            26.9265767
mois^2        -123.7120431
R.squared        0.9515476
n                5.0000000
critical.mois    0.1088276
max.bulk         1.9477727
```
![](<man/figures/proctor.jpeg>)

## Soil water availability tools

Quantifying the soil water availability for plants through the IWC approach:

```r

# Usage
iwc(theta_R = 0.166, theta_S = 0.569, alpha = 0.029, n = 1.308, 
    a = 0.203, b = 0.256, hos = 200, graph = TRUE)
 
# Results
                     IWC       EI         h.Range
EKa(h, hos)    0.0144000   0.9600  66.43 - 139.49
EK(h, hos)     0.0405000   5.3700    139.49 - 330
C(h, hos)      0.0846000  49.4800   330 - 2471.44
ER(h, hos)     0.0288000  87.1200 2471.44 - 15000
ERKdry(h, hos) 0.0006000   4.9100   12000 - 15000
Sum            0.1689139 147.8336       0 - 15000

```

Quantifying the soil water availability for plants through the LLWR approach:

```r
# Usage
data(skp1994)
ex1 <- with(skp1994,
	 llwr(theta = W, h = h, Bd = BD, Pr = PR,
		particle.density = 2.65, air = 0.1,
		critical.PR = 2, h.FC = 100, h.WP = 15000))
```		
![](https://github.com/arsilva87/soilphysics/blob/master/man/figures/LLWR.jpg)
![](<man/figures/LLWR.jpg>)

Quantifying the LLWR using van Genuchten's parameters:
```r
# Usage
  par(mfrow=c(1,2))
  llwr_llmpr(thetaR=0.1180, thetaS=0.36, alpha=0.133, n=1.30, 
        d=0.005, e=-2.93, f=3.54, PD=2.65,
        critical.PR=4, h.FC=100, h.PWP=15000, air.porosity=0.1,
        labels=c("AFP", "FC","PWP", "PR"),
        graph1=TRUE,graph2=FALSE, ylab=expression(LLMPR~(hPa)), ylim=c(15000,1))
  mtext(expression("Bulk density"~(Mg~m^-3)),1,line=2.2, cex=0.8)

  llwr_llmpr(thetaR=0.1180, thetaS=0.36, alpha=0.133, n=1.30, 
        d=0.005, e=-2.93, f=3.54, PD=2.65,
        critical.PR=4, h.FC=100, h.PWP=15000, air.porosity=0.1,
        labels=c("AFP", "FC","PWP", "PR"),
        graph1=FALSE,graph2=TRUE, ylab=expression(LLMPR~(hPa)), ylim=c(0.1,0.5))
  mtext(expression("Bulk density"~(Mg~m^-3)),1,line=2.2, cex=0.8)

# Results
$CRITICAL_LIMITS
     theta potential
AIR 0.2600     41.02
FC  0.2285    100.00
PWP 0.1428  15000.00
PR  0.1939    356.65

$LLRW_LLMPR
         Upper    Lower    Range
LLWR    0.2285   0.1939   0.0346
LLMPR 100.0000 356.6500 256.6500

```
![](<man/figures/LLWR2.jpeg>)


## Precompression stress

Estimating the precompression stress by several methods:
![](<man/figures/sigmaPfluxo.jpg>)
```r
pres <- c(1, 12.5, 25, 50, 100, 200, 400, 800, 1600)
VR <- c(0.846, 0.829, 0.820, 0.802, 0.767, 0.717, 0.660, 0.595, 0.532)

# Usage
sigmaP(VR, pres, method = "casagrande", n4VCL = 2)

# Results
Preconsolidation stress: 104.2536
Method: casagrande, with mcp equal to 1.7885
Compression index: 0.2093
Swelling index: 0.0179 
```

## Soil water retention curve

Fitting water retention curve using van Genuchten's model

```r
h <- c(0.001, 50.65, 293.77, 790.14, 992.74, 5065, 10130, 15195)
w <- c(0.5650, 0.4013, 0.2502, 0.2324, 0.2307, 0.1926, 0.1812, 0.1730)

# Usage
fitsoilwater(theta=w, x=h, ylim=c(0.1,0.6))

# Results
Parameters:
        Estimate Std. Error t value Pr(>|t|)    
theta_R  0.16761    0.01272  13.179 0.000192 ***
theta_S  0.56531    0.01092  51.786 8.32e-07 ***
alpha    0.04748    0.01177   4.035 0.015671 *  
n        1.52926    0.09579  15.965 9.00e-05 ***
---
```
![](<man/figures/wrc.jpg>)

## S Index
```r
# Usage
Sindex(theta_R=0, theta_S=0.395, alpha=0.0217, 
       n=1.103)
       
# Results
          The S Index 

h_i : 395.4757 
theta_i : 0.3139 
|S| : 0.0296 
Soil physical quality : Poor 
```

## Soil Aggregate-Size Distribution
```r
data(SoilAggregate)
head(SoilAggregate)
  ID    D3 D1.5 D0.75 D0.375 D0.178 D0.053
1 A1 25.80 7.55  5.50   5.10   3.00   3.05
2 A2 19.85 5.30  7.45   7.30   4.40   5.70
3 A3  7.10 9.80 11.60   8.10   2.35  11.05
4 B1  6.10 4.85 11.20  13.10   7.15   7.60
5 B2 12.00 6.30 16.10   7.35   3.70   4.55
6 B3 14.10 6.15  8.80  11.05   4.60   5.30

classes <- c(3, 1.5, 0.75, 0.375, 0.178, 0.053)

# Usage
out <- aggreg.stability(sample.id = SoilAggregate[ ,1], 
                        dm.classes = classes, 
                        aggre.mass = SoilAggregate[ ,-1])

		
# Results
head(out)
  sample.id      MWD       GMD total.mass X3 X1.5 X0.75 X0.375 X0.178 X0.053
1        A1 1.909163 1.2382214         50 52   15    11     10      6      6
2        A2 1.538206 0.8239103         50 40   11    15     15      9     11
3        A3 0.974829 0.4865272         50 14   20    23     16      5     22
4        B1 0.811260 0.4311214         50 12   10    22     26     14     15
5        B2 1.223620 0.7282644         50 24   13    32     15      7      9
6        B3 1.267369 0.6853162         50 28   12    18     22      9     11
```


# Spin-off

Sedimentation time of soil particle (Stokes' law): <https://renatoagro.shinyapps.io/stokesapp/>

Exploring water retention curve using van Genuchten's model: <https://soilphysics.shinyapps.io/wrcAPP/>

Soil Aggregate-Size Distribution: <https://renatoagro.shinyapps.io/Agre/>

Usual Least Limiting Water Range (LLWR): <https://soilphysics.shinyapps.io/LLWRAPP/>

Water suction at the point of hydraulic cut-off (Dexter et al. 2012): <https://soilphysics.shinyapps.io/h_cutoff/>

LLWR and LLMPR:  <https://soilphysics.shinyapps.io/LLWR_LLMPR/>

PredComp 1.0:  <https://appsoilphysics.shinyapps.io/PredComp/>

fisoilwater:  <https://appsoilphysics.shinyapps.io/fitsoilwaterAPP/>

Automatically fit the Water Retention Curve online: [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gist/arsilva87/59d2a78666cf13ab8fb60bafaf7e65b8/HEAD?urlpath=shiny)

 
# Citation and references

De Lima, R.P.; Tormena, C.A.; Figueiredo, G.C; Da Silva, A.R.; Rolim, M.M. (2020) Least limiting water and matric potential ranges of agricultural soils with calculated physical restriction thresholds. *Agricultural Water Management*, 240: 106299. DOI: <https://doi.org/10.1016/j.agwat.2020.106299>

Da Silva, A.R.; De Lima, R.P. (2017) Determination of maximum curvature point with the R package soilphysics. *International Journal of Current Research*, 9: 45241-45245.  

De Lima, R.P.; Da Silva, A.R.; Da Silva, A.P.; Leao, T.P.; Mosaddeghi, M.R. (2016) soilphysics: an R package for calculating soil water availability to plants by different soil physical indices. *Computers and Eletronics in Agriculture*, 120: 63-71. DOI: <https://doi.org/10.1016/j.compag.2015.11.003>

Da Silva, A.R.; De Lima, R.P. (2015) soilphysics: an R package to determine soil preconsolidation pressure. *Computers and Geosciences*, 84: 54-60. DOI: <https://doi.org/10.1016/j.cageo.2015.08.008>


# Contributions and bug reports

soilphysics is an ongoing project. Then, contributions are very welcome. If you have a question or have found a bug, please open an [Issue](https://github.com/arsilva87/soilphysics/issues) or reach out directly by e-mail: <anderson.silva@ifgoiano.edu.br> or <renato_agro_@hotmail.com>.

# By
Anderson R. da Silva & Renato P. de Lima


NEWS
------------------------------------------------
                version 5.0

* New shiny function (WRC_App) to automatically fit the water retention curves
* Fixed on documentation

------------------------------------------------
                version 4.1

* Documentation fixes
* Starting message fixed

------------------------------------------------
                version 4.0

* Documentation fixes
* Code fixes in soilStrength
* New functions: particle.sedimentation(), aggreg.stability(), r(), 
  psd(), Kr(), llwr_llmpr(), hydraulicCutOff2()
* Several shiny apps
* New data: SoilAggregate
* Fixes on tkrplot and rpanel dependency
* New citation entry

------------------------------------------------
                version 3.1

* Documentation fixes
* New functions: stressTraffic(), for predicting the 
  applied stress in the soil profile; soilStrength(),
  for predicting load bearing capacity; llwrPTF() and soilDeformation()
* Output "LLWR" of llwr() fixed
* New reference for citation()

------------------------------------------------
                version 3.0

* Documentation fixes
* New references for citation()
* Data skp1994 were updated (now in full)

------------------------------------------------
                version 2.4

* Documentation fixes
* New arguments for llwr() and new graphical output
* Missing legends added to iwc() graphs

------------------------------------------------
                version 2.3

* Documentation fixes
* New detailed output for iwc() and new
	graphical arguments
* Three new methods for calculating 'mcp' with maxcurv()

------------------------------------------------
                version 2.2

* Documentation fixes
* New argument (hos = 0) on iwc() for dealing with
	salinity effects
* Boundaries for the argument mcp of sigmaP()

------------------------------------------------
                version 2.1

* Documentation fixes
* corretions on the graphic output of iwc()
* 'mcp' bug fixed on sigmaP()
* new output for sigmaP(), now with compression and 
	decompression indices
* print method for sigmaP()
* new functions: simSigmaP() and plotCIsigmaP()
* bug fixed on maxcurv()
* new graphical output for maxcurv()

------------------------------------------------
                version 2.0

* Documentation fixes
* new argument (water.model) in the function llwr()
* new graphic output and print.method for llwr()
* R-squared for the models fitted in llwr()
* R-squared for the models fitted in fitsoilwater(), 
	fitsoilwater2(), fitsoilwater3(), fitsoilwater5()
* new functions: soilwater2(), fitsoilwater2(), 
	soilwater3(), fitsoilwater3(), soilwater4(), 
	fitsoilwater4(), soilwater5(), fitsoilwater5(), 
	soilwater6(), fitsoilwater6(), hydraulicCutOff(),
	hemc(), iwc()

------------------------------------------------
                version 1.2

* Documentation fixes
* function llwr()
* data set skp1994
* automatic casagrande method in the function sigmaP()
* a flow chart in the documentation of sigmaP()

------------------------------------------------
                version 1.1

* Documentation fixes
* S3 print method for the classes: criticalmoisture, 
	maxcurv and S.index

------------------------------------------------
                version 1.0

* First version released on CRAN


R/WRC_App.R
findWRC = function(w, h,
                   mod2fit = c(TRUE, TRUE, TRUE, TRUE, TRUE))
{
  y = w; x = h
  tS = max(y)
  tR = min(y)
  stopifnot(any(mod2fit == TRUE))

  # Eqs
  models = c("van Genuchten", "Brooks & Corey",
             "Groenevelt & Grant", "Dexter", "FXW")
  npar = c(2, 2, 3, 4, 3)
  eqVG = expression( tR + (tS - tR)*(1 + (a * x)^n)^(-(1 - 1/n)) )
  eqBC = expression( ifelse(x < hb, tS, tR + (tS - tR)*(x/hb)^-lambda) )
  eqGG = expression( k1 * (exp(-k0/6.653^n) - exp(-k0/x^n)) )
  eqDE = expression( tR + a1 * exp(-x/p1) + a2 * exp(-x/p2) )
  eqFX = expression( tS * (1 - log(1 + x/1500)/log(1 + 6300000/1500)) *
                       (log(exp(1) + abs(a * x)^n))^-m )
  exps = c(eqVG, eqBC, eqGG, eqDE, eqFX)

  # -logliks
  nllVG = function(par) {
    e = y - eval(eqVG, list(a = par[1], n = par[2]))
    -sum(dnorm(e, 0, sd(e), log = T))
  }
  nllBC = function(par) {
    e = y - eval(eqBC, list(hb = par[1], lambda = par[2]))
    -sum(dnorm(e, 0, sd(e), log = T))
  }
  nllGG = function(par) {
    e = y - eval(eqGG, list(k0 = par[1], k1 = par[2], n = par[3]))
    -sum(dnorm(e, 0, sd(e), log = T))
  }
  nllDE = function(par) {
    e = y - eval(eqDE, list(a1 = par[1], p1 = par[2],
                            a2 = par[3], p2 = par[4]))
    -sum(dnorm(e, 0, sd(e), log = T))
  }
  nllFX = function(par) {
    e = y - eval(eqFX, list(a = par[1], n = par[2], m = par[3]))
    -sum(dnorm(e, 0, sd(e), log = T))
  }


  # starting values
  if (mod2fit[1]) {
    grVG = expand.grid(a = seq(0.002, 0.2, len = 30),
                       n = seq(1, 4, len = 30))
    criVG = apply(grVG, 1, nllVG)
    oVG = which.min(criVG)
    mleVG = optim(par = grVG[oVG, ], nllVG, hessian = TRUE,
                  control = list(maxit = 1000))
  }
  if (mod2fit[2]) {
    grBC = expand.grid(hb = seq(0, 100, len = 50),
                       lambda = seq(0.1, 3, len = 30))
    criBC = apply(grBC, 1, nllBC)
    oBC = which.min(criBC)
    mleBC = optim(par = grBC[oBC, ], nllBC, hessian = TRUE,
                  control = list(maxit = 1000))
  }
  if (mod2fit[3]) {
    grGG = expand.grid(k0 = seq(1, 20, len = 30),
                       k1 = seq(0, 1, len = 30),
                       n = seq(1, 10, len = 20))
    criGG = apply(grGG, 1, nllGG)
    oGG = which.min(criGG)
    mleGG = optim(par = grGG[oGG, ], nllGG, hessian = TRUE,
                  control = list(maxit = 1000))
  }
  if (mod2fit[4]) {
    grDE = expand.grid(a1 = seq(0.01, 0.4, len = 12),
                       p1 = seq(500, 10000, len = 20),
                       a2 = seq(0.01, 0.5, len = 12),
                       p2 = seq(1, 2000, len = 12))
    criDE = apply(grDE, 1, nllDE)
    oDE = which.min(criDE)
    mleDE = optim(par = grDE[oDE, ], nllDE, hessian = TRUE,
                  control = list(maxit = 1000))
  }
  if (mod2fit[5]) {
    grFX = expand.grid(a = seq(0.002, 0.2, len = 30),
                       n = seq(1, 4, len = 30),
                       m = seq(0, 1, len = 20))
    criFX = apply(grFX, 1, nllFX)
    oFX = which.min(criFX)
    mleFX = optim(par = grFX[oFX, ], nllFX, hessian = TRUE,
                  control = list(maxit = 1000))
  }
  mles = list(VG = if(mod2fit[1]) mleVG else NA,
              BC = if(mod2fit[2]) mleBC else NA,
              GG = if(mod2fit[3]) mleGG else NA,
              DE = if(mod2fit[4]) mleDE else NA,
              FX = if(mod2fit[5]) mleFX else NA)

  pars = list(VG = if(mod2fit[1]) mleVG$par else NA,
              BC = if(mod2fit[2]) mleBC$par else NA,
              GG = if(mod2fit[3]) mleGG$par else NA,
              DE = if(mod2fit[4]) mleDE$par else NA,
              FX = if(mod2fit[5]) mleFX$par else NA)

  # Fitting criteria
  aic = c(VG = if(mod2fit[1]) 2*mleVG$value + 2*2 else NA,
          BC = if(mod2fit[2]) 2*mleBC$value + 2*2 else NA,
          GG = if(mod2fit[3]) 2*mleGG$value + 2*3 else NA,
          DE = if(mod2fit[4]) 2*mleDE$value + 2*4 else NA,
          FX = if(mod2fit[5]) 2*mleFX$value + 2*3 else NA
  )
  std = list(VG = if(mod2fit[1]) sqrt(diag(solve(mleVG$hessian))) else NA,
             BC = if(mod2fit[2]) sqrt(diag(solve(mleBC$hessian))) else NA,
             GG = if(mod2fit[3]) sqrt(diag(solve(mleGG$hessian))) else NA,
             DE = if(mod2fit[4]) sqrt(diag(solve(mleDE$hessian))) else NA,
             FX = if(mod2fit[5]) sqrt(diag(solve(mleFX$hessian))) else NA
  )
  eVG = if(mod2fit[1]) y - eval(eqVG, as.list(mleVG$par)) else NA
  eBC = if(mod2fit[2]) y - eval(eqBC, as.list(mleBC$par)) else NA
  eGG = if(mod2fit[3]) y - eval(eqGG, as.list(mleGG$par)) else NA
  eDE = if(mod2fit[4]) y - eval(eqDE, as.list(mleDE$par)) else NA
  eFX = if(mod2fit[5]) y - eval(eqFX, as.list(mleFX$par)) else NA

  r2 = c(VG = if(mod2fit[1]) 1 - sum(eVG^2)/sum((y - mean(y))^2) else NA,
         BC = if(mod2fit[2]) 1 - sum(eBC^2)/sum((y - mean(y))^2) else NA,
         GG = if(mod2fit[3]) 1 - sum(eGG^2)/sum((y - mean(y))^2) else NA,
         DE = if(mod2fit[4]) 1 - sum(eDE^2)/sum((y - mean(y))^2) else NA,
         FX = if(mod2fit[5]) 1 - sum(eFX^2)/sum((y - mean(y))^2) else NA
  )
  mape = c(VG = if(mod2fit[1]) 100*mean(abs(eVG)/y) else NA,
           BC = if(mod2fit[2]) 100*mean(abs(eBC)/y) else NA,
           GG = if(mod2fit[3]) 100*mean(abs(eGG)/y) else NA,
           DE = if(mod2fit[4]) 100*mean(abs(eDE)/y) else NA,
           FX = if(mod2fit[5]) 100*mean(abs(eFX)/y) else NA
  )

  # out
  dtf = data.frame(Rsq = r2, MAPE = mape, AIC = aic, npar)
  best_id = which.min(dtf[, "AIC"])
  best = paste("Best fit (AIC):", models[best_id])

  w_range = c(tR = tR, tS = tS)
  pa <- c(pars[[best_id]], w_range)
  pas <- as.list(pa)
  eq <- exps[best_id]
  for(j in 1:length(pas)) {
    eq <- gsub(names(pas)[j], round(pas[[j]], 5), eq, fixed = TRUE)
  }

  out = list(e = exps[best_id],
             fitting_criteria = dtf[complete.cases(dtf), ],
             best_fit = models[best_id], equation = eq,
             par = pars[[best_id]], std = std[[best_id]],
             w_range = w_range)
  return(out)
}



#----
about <- fluidPage(
         withMathJax(),
         h6("Model equations:"),
         h6('- van Genuchten (VG)
            $$\\theta (m^3 m^{-3}) = \\theta_R + (\\theta_S - \\theta_R)(1 + (\\alpha x)^n)^{1/n -1}$$'),
         h6('- Brooks & Corey (BC)
            $$\\theta (m^3 m^{-3}) = \\theta_R + (\\theta_S - \\theta_R)(x_b/x)^{\\lambda}$$'),
         h6('- Groenevelt & Grant (GG)
            $$\\theta (m^3 m^{-3}) = k_1 \\exp(-k_0 / x_0^n) - k_1 \\exp(-k_0 / x^n)$$'),
         h6('- Dexter (DE)
            $$\\theta (m^3 m^{-3}) = \\theta_R + a_1 \\exp(-x/p_1) + a_2 \\exp(-x/p_2)$$'),
         h6('- FXW (Fredlung-Xing-Wang, Fredlung and Xing 1994; Wang et al. 2018)
            $$\\theta (m^3 m^{-3}) = \\theta_S \\left(1 - \\frac{\\log(1 + x/h_r)}{\\log(1 + h_0/h_r)}\\right)
                      [\\log(\\exp(1) + |a x|^n)]^{-m}$$'),
         h6('where $$h_r = 1500, h_0 = 6.3 \\times 10^6$$ cm
            (Schneider and Goss, 2012; Rudiyanto et al., 2021)'),
         h6("Please check the documentation of the functions 'soilwater'
            of package soilphysics to know more about each equation."),
                  h6("Statistics legends:
                                      Rsq: (pseudo) coefficient of determination,
                                      MAPE: mean absolute percentual error,
                                      AIC: Akaike information criterion,
                                      npar: number of estimated parameters,
                                      std: standard deviation"),
                  h6("The parameters are estimated using the maximum
                                      likelihood method, based upon the normal
                                      distribution of model residuals. For optimization
                                      purposes, the values of the residual (theta_R)
                                      and saturation (theta_S) water content are
                                      not estimated. Instead, the minimum and maximum values
                                      are returned in the output 'w_range'"),
                  h6("This app can run online: ",
                     tags$a(href = "https://mybinder.org/v2/gist/arsilva87/59d2a78666cf13ab8fb60bafaf7e65b8/HEAD?urlpath=shiny",
                            "click here")),
                  h6("The fitsoilwater app offers another alternative to
                                      interactively fit water retention curves.",
                     tags$a(href = "https://appsoilphysics.shinyapps.io/fitsoilwaterAPP/",
                            "https://appsoilphysics.shinyapps.io/fitsoilwaterAPP/")),
                  h6("Citation:
                                      da Silva, A. R., de Lima, R. P. (2022) soilphysics:
                                      Basic and Model-Based Soil Physical Analyses.
                                      R package version 5.0.",
                     tags$a(href = "https://cran.r-project.org/package=soilphysics",
                            "https://cran.r-project.org/package=soilphysics")),
                  h6("Find out more about the soilphysics project: ",
                     tags$a(href = "https://arsilva87.github.io/soilphysics/",
                            "https://arsilva87.github.io/soilphysics/"))
)


# ---------------------------------------------------------------------
ui_findWRC <- dashboardPage(
  dashboardHeader(title = "WRC App: Maximum Likelihood fit of Water Retention Curve",
                  titleWidth = 600),
  dashboardSidebar(
    h6(tags$a(href = "https://arsilva87.github.io/soilphysics/",
              "Powered by: soilphysics")),
    sidebarMenu(
       menuItem("Import data", icon = icon("folder-open"),
                h6(tags$a(href="https://ce99d4d6-d4c5-48a3-b911-9e83247054ca.filesusr.com/ugd/45a659_170e68a8ee1f474b9493bc67df713eed.csv?dn=MyWRCExample.csv",
                          "(example)")),
                    fileInput("file", "(comma-separated values)",
                              accept=c('text/csv',
                                       'text/comma-separated-values,text/plain',
                                       '.csv')
                    ),
                checkboxInput("semicolon", "Semicolon-separated"),
                checkboxInput("dec", "Comma as decimal")
       ),
       menuItem("Select columns", icon = icon("table"),
                selectInput('w', 'Water content (y)', ''),
                selectInput('h', 'Matric potential (x)', ''),
                checkboxInput("logh", "Log10( x )")
       ),
       menuItem("Models to fit", icon = icon("check"),
                checkboxInput("modVG", "van Genuchten", value = TRUE),
                checkboxInput("modBC", "Brooks & Corey", value = TRUE),
                checkboxInput("modGG", "Groenevelt & Grant", value = TRUE),
                checkboxInput("modDE", "Dexter", value = TRUE),
                checkboxInput("modFX", "FXW", value = TRUE)
       ),
       actionButton("run", "RUN", width = "40%",
                    icon = icon("r-project"))
    )
  ),
  dashboardBody(
    tabsetPanel(
      tabPanel(icon("home"),
        fluidRow(
          box(
            title = "Data", status = "warning", solidHeader = TRUE,
            collapsible = TRUE, width = 12,
            h5("Input data (editable, left-click)
               (right-click to insert or remove rows)"),
            rHandsontableOutput("tab")
          ),
          tabBox(title = "Results", width = 12,
            tabPanel("Statistics", icon = icon("calculator"),
              verbatimTextOutput("fit")
            ),
            tabPanel("Best-fit", icon = icon("chart-line"),
              plotOutput("graph", height = 480, width = 650)
            )
          )
        )
      ),
      tabPanel(icon("info"), about)
    )
  )
)



# ---------------------------------------------------
server_findWRC <- function(session, input, output) {
  dataIn <- reactive( req(input$file) )
  output$tab <- renderRHandsontable({
    dec <- ifelse(input$dec, ",", ".")
    if(input$semicolon) {
       DF <- read.csv2(dataIn()$datapath, dec = dec)
    } else {
       DF <- read.csv(dataIn()$datapath, dec = dec)
    }
    updateSelectInput(session, inputId = 'w',
                      choices = names(DF), selected = NULL)
    updateSelectInput(session, inputId = 'h',
                      choices = names(DF), selected = NULL)
    rhandsontable(DF)
  })
  mt <- reactive({
    DF = hot_to_r(input$tab)
    if(!is.null(input$tab$changes$changes)) {
      row.no <- unlist(input$tab$changes$changes)[1]
      col.no <- unlist(input$tab$changes$changes)[2]
      new.val <- unlist(input$tab$changes$changes)[4]
    }
    DF
  })
  observeEvent(input$run, {
    xy <- mt()[, c(input$h, input$w)]
    if (input$logh) {
      xy[,1] <- log10(xy[,1])
    }
    r <- findWRC(w = xy[,2], h = xy[,1],
                 mod2fit = c(input$modVG, input$modBC,
                             input$modGG, input$modDE, input$modFX))
    output$fit <- renderPrint( r[-1] )
    output$graph <- renderPlot({
      par(mar = c(4.5, 4.5, 2.5, 1), las = 1, cex = 1.3)
      plot(xy, cex = 2, log = "x")
      pa <- c(r$par, r$w_range)
      pars <- as.list(pa)
      eq <- r$e
      f <- function(x) eval(eq, pars)
      curve(f, add = TRUE, col = "steelblue3", lwd = 4)
      for(j in 1:length(pars)) {
        eq <- gsub(names(pars)[j], round(pars[[j]], 4), eq, fixed = TRUE)
      }
      leg <- as.expression(paste0("y = ", eq))
      title(leg, cex.main = 0.9)
    })
  })
}

# Run the app
WRC_App <- function() {
  shinyApp(ui_findWRC , server_findWRC)
  }
